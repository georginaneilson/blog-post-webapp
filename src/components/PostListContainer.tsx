import { usePosts } from "../hooks/usePosts";
import Post, { PostItem } from "./Post";
import SearchBar from "./SearchBar";
import { useState } from "react";

const ListContainer = () => {
  const [searchValue, setSearchValue] = useState("");
  const { data: posts, handleDelete, deleteError } = usePosts();

  const handleSearch = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchValue(e.target.value);
  };

  const filteredPostArray = (posts: PostItem[], searchTerm: string) => {
    const filteredPosts = posts.filter((post) => {
      return post.title.includes(searchTerm);
    });

    return filteredPosts.map((post) => (
      <Post
        key={post.id}
        post={post}
        handleDelete={handleDelete}
        error={deleteError === post.id}
      />
    ));
  };

  const postArray = () => {
    return posts.map((post: PostItem) => (
      <Post
        key={post.id}
        post={post}
        handleDelete={handleDelete}
        error={deleteError === post.id}
      />
    ));
  };

  return (
    <div>
      <SearchBar handleSearch={handleSearch} searchValue={searchValue} />
      <div className="mx-4 md:mx-50 md:px-2 xl:mx-100 md:h-screen overflow-y-auto mt-4">
        {searchValue === ""
          ? postArray()
          : filteredPostArray(posts, searchValue)}
      </div>
    </div>
  );
};

export default ListContainer;
