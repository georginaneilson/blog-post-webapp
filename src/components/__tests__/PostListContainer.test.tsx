import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import PostListContainer from "../PostListContainer";
import { act } from "react";

// Mock data
const mockPosts = [
  {
    id: 1,
    title: "First Post",
    body: "This is the first post",
    userId: 1,
  },
  {
    id: 2,
    title: "Second Post",
    body: "Another one",
    userId: 2,
  },
];

let mockDeletePostData = 200;

// Mock API functions
// TODO: in future rather than mock the api, we should use a mock server like MSW
vi.mock("../../services/jsonPlaceholder", () => ({
  getPostsData: () => Promise.resolve(mockPosts),
  deletePostData: () => mockDeletePostData,
}));

describe("PostListContainer", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("matches the snapshot", () => {
    const result = render(<PostListContainer />);
    expect(result).toMatchSnapshot();
  });

  it("fetches and renders posts", async () => {
    render(<PostListContainer />);

    expect(screen.getByPlaceholderText(/search/i)).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.getByText("First Post")).toBeInTheDocument();
      expect(screen.getByText("Second Post")).toBeInTheDocument();
    });
  });

  it("filters posts based on search input", async () => {
    render(<PostListContainer />);

    const searchInput = screen.getByPlaceholderText(/search/i);

    await act(async () => {
      await userEvent.type(searchInput, "First");
    });

    await waitFor(() => {
      expect(screen.getByText("First Post")).toBeInTheDocument();
      expect(screen.queryByText("Second Post")).not.toBeInTheDocument();
    });
  });

  it("deletes a post and updates the UI", async () => {
    render(<PostListContainer />);

    await waitFor(() => {
      expect(screen.getByText("First Post")).toBeInTheDocument();
      expect(screen.getByText("Second Post")).toBeInTheDocument();
    });

    const deleteButtons = screen.getAllByRole("button", {
      name: /delete post/i,
    });

    await act(async () => {
      await userEvent.click(deleteButtons[0]);
    });
    await waitFor(() => {
      expect(screen.queryByText("First Post")).not.toBeInTheDocument();
      expect(screen.getByText("Second Post")).toBeInTheDocument();
    });
  });

  it("displays an error when the delete fails", async () => {
    mockDeletePostData = 404;
    render(<PostListContainer />);

    await waitFor(() => {
      expect(screen.getByText("First Post")).toBeInTheDocument();

      expect(
        screen.queryByText("Error deleting post, please try again")
      ).not.toBeInTheDocument();
    });

    const deleteButtons = screen.getAllByRole("button", {
      name: /delete post/i,
    });

    await act(async () => {
      await userEvent.click(deleteButtons[0]);
    });
    await waitFor(() => {
      expect(screen.getByText("First Post")).toBeInTheDocument();
      expect(
        screen.getByText("Error deleting post, please try again")
      ).toBeInTheDocument();
    });
  });
});
