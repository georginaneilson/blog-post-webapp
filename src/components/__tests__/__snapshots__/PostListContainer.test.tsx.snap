// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostListContainer > matches the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="h-[65px] bg-[#ffe1e6] md:bg-white p-4 sticky top-0 z-10 "
        >
          <h1
            class="sr-only"
          >
            Blog Posts
          </h1>
          <span
            class="flex items-center gap-2 bg-white md:bg-[#fff7f9] rounded-xl p-2 md:mx-50 xl:mx-100 focus-within:outline-2 focus-within:outline-pink-700 shadow-sm"
          >
            <label
              class="sr-only"
              for="search"
            >
              Search blog posts
            </label>
            <input
              class="w-full h-full bg-white md:bg-[#fff7f9] placeholder:font-medium placeholder:!text-gray-400 font-medium outline-none "
              data-testid="search-test"
              id="search"
              placeholder="Search blog posts..."
              title="Search blog posts"
              type="text"
              value=""
            />
            <svg
              class="w-4 h-4 fill-gray-400"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
              />
            </svg>
          </span>
        </div>
        <div
          class="mx-4 md:mx-50 md:px-2 xl:mx-100 md:h-screen overflow-y-auto mt-4"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="h-[65px] bg-[#ffe1e6] md:bg-white p-4 sticky top-0 z-10 "
      >
        <h1
          class="sr-only"
        >
          Blog Posts
        </h1>
        <span
          class="flex items-center gap-2 bg-white md:bg-[#fff7f9] rounded-xl p-2 md:mx-50 xl:mx-100 focus-within:outline-2 focus-within:outline-pink-700 shadow-sm"
        >
          <label
            class="sr-only"
            for="search"
          >
            Search blog posts
          </label>
          <input
            class="w-full h-full bg-white md:bg-[#fff7f9] placeholder:font-medium placeholder:!text-gray-400 font-medium outline-none "
            data-testid="search-test"
            id="search"
            placeholder="Search blog posts..."
            title="Search blog posts"
            type="text"
            value=""
          />
          <svg
            class="w-4 h-4 fill-gray-400"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
            />
          </svg>
        </span>
      </div>
      <div
        class="mx-4 md:mx-50 md:px-2 xl:mx-100 md:h-screen overflow-y-auto mt-4"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
